src/store/options.js

```diff
@@ -75,14 +71,38 @@ const Options = {
 
 export default Options;
 
-chrome.runtime.onMessage.addListener((message) => {
-  if (isUpdateOptionsMessage(message)) {
+const observers = new Set();
+export async function observe(property, fn) {
+  let value;
+  const wrapperFn = (options) => {
+    if (value === undefined || options[property] !== value) {
+      value = options[property];
+      fn(value);
+    }
+  };
+
+  observers.add(wrapperFn);
+
+  // Run for the first time with the current options
+  const options = await store.resolve(store.get(Options));
+  wrapperFn(options);
+
+  // Return unobserve function
+  return () => {
+    observers.delete(wrapperFn);
+  };
+}
+
+chrome.runtime.onMessage.addListener((msg) => {
+  if (msg.action === UPDATE_OPTIONS_ACTION_NAME) {
     const options = store.get(Options);
 
     if (!store.pending(options)) {
       store.clear(options, false);
       store.get(Options);
     }
+
+    observers.forEach((observer) => observer(msg.options));
```

Comment:
all those calls should be wrapped with try/catch

Reply:
Is there a particular reason to do that here? We were even throwing (in old code in another place) inside of the `addListener`, but without try/catch.

Reply:
I see the argument for always wrapping fire-and-forget calls of async-functions. The main argument IMHO is that proper error messages can be useful, especially once you hit the nasty bugs that are hard to reproduce.

Here you have operations like local storage in the chain, which can fail in very rare situations. Especially in places like here where the async operations should practically never fail, logging the rare cases when they do fail can be useful when tracking down obscure bugs later.

Reply:
Ok, now i get the context, it is about using `await`. Got it.

Reply:
I looked at other examples of async functions in code, and usually, there is no try/catch. And it is right. The async function which has `await` which rejects, simply returns a rejected promise with an error. If the consumer of the function does not catch that (in another await or promise chain) it is always displayed in the console with the error message with proper stack trace, message, and place from the file. I think we should use `try/catch` if we want to **do something** in that catch block (like clean up things because of the error). I don't like putting those `console.log` just to notify something, which would be notified whatsoever.

I am not sure if the following change adds something meaningful to the code:

```js
  try {
    // Run for the first time with the current options
    const options = await store.resolve(store.get(Options));
    wrapperFn(options);
  } catch (e) {
    console.error(e);
    throw e;
  }
```

Reply:
maybe this is a better place to have try/catch https://github.com/ghostery/ghostery-dnr-extension/pull/114/files#diff-b200b671a2beeb20552b88eef51aa781388bd9434a0742efe2d48769166479cfR80

the point is: when one callback throws, the others should still fire.

Reply:
Well, if skipping calls is the concern, then this should happen here. Since there are async calls, all callbacks will be executed (unless I'm missing something).

In NodeJs, I get the feeling that it became best practice to avoid unrejected promises. In the browsers, it is not clear if that is the case. For me, it is most about seeing errors. Sometimes, explicit error messages can help to provide context (might be possible here). The usefulness of stack-trace in JavaScript sometimes depends on whether the browsers can fill in the async call chain.

But from the functionality, I think the code is correct as is.